File: generator-common/src/main/java/io/github/potjerodekool/openapi/common/util/CollectionBuilder.java
package io.github.potjerodekool.openapi.common.util;

import java.util.*;
import java.util.function.Function;

public class CollectionBuilder<T> {

    private final List<T> list = new ArrayList<>();

    public CollectionBuilder() {
    }

    public CollectionBuilder(final List<T> values) {
        this.list.addAll(values);
    }

    public CollectionBuilder<T> add(final T value) {
        this.list.add(value);
        return this;
    }

    public CollectionBuilder<T> addAll(final Collection<T> values) {
        this.list.addAll(values);
        return this;
    }

    public <R> CollectionBuilder<R> map(final Function<T, R> mapper) {
        final var mappedValues = this.list.stream()
                .map(mapper)
                .toList();
        return new CollectionBuilder<>(mappedValues);
    }

    public List<T> buildList() {
        return list;
    }

    public Set<T> buildSet() {
        return new HashSet<>(list);
    }
}

File: generator-common/src/main/kotlin/io/github/potjerodekool/openapi/common/util/CollectionBuilder.kt
package io.github.potjerodekool.openapi.common.util

import java.util.function.Function

class CollectionBuilder<T> {
    private val list: MutableList<T> = ArrayList()

    constructor()

    constructor(values: List<T>?) {
        list.addAll(values!!)
    }

    fun add(value: T): CollectionBuilder<T> {
        list.add(value)
        return this
    }

    fun addAll(values: Collection<T>?): CollectionBuilder<T> {
        list.addAll(values!!)
        return this
    }

    fun <R> map(mapper: Function<T, R>?): CollectionBuilder<R> {
        val mappedValues = list.stream()
            .map(mapper)
            .toList()
        return CollectionBuilder(mappedValues)
    }

    fun buildList(): List<T> {
        return list
    }

    fun buildSet(): Set<T> {
        return HashSet(list)
    }
}
